### DISCUSSION TITLE
Conduct an Audit

### DISCUSSION THREAD

**Discussion by Dadadah** [noisetorch/NoiseTorch#264] [General]
I assume you have regained control of your security keys.

I think this project is large enough to fund an open source audit to alleviate any concerns about security compromises.

These things can happen and there are things the repository owners can do to address it.

Initial research shows an audit taking about US$100/1000 lines of code.

Perhaps consider starting a fundraiser for it @lawl ?

**Comment by lawl** [comment]
No i havent. I believe its unrelated to noisetorch, but better safe than sorry. I am dead right now and dont have the energy to rebuild my computing devices from scratch.

Everythings still compromised, but hopefully enouch people know now.

**Comment by OldiLo** [comment]
Should I remove NoiseTorch and ayy project?

**Comment by CheaterTim** [comment]
@OldiLo 
> Should I remove NoiseTorch and ayy project?

I would remove both for now and watch out for any updates regarding the situation

**Comment by OldiLo** [comment]
> I assume you have regained control of your security keys.
> 
> I think this project is large enough to fund an open source audit to alleviate any concerns about security compromises.
> 
> These things can happen and there are things the repository owners can do to address it.
> 
> Initial research shows an audit taking about US$100/1000 lines of code.
> 
> Perhaps consider starting a fundraiser for it @lawl ?

Maybe the community should do it

**Comment by lawl** [comment]
No, dont give money. Patreon is scheduled for deletion cant do it instantly.

**Comment by OldiLo** [comment]
> No, dont give money. Patreon is scheduled for deletion cant do it instantly.

No, the community can do an audit

**Comment by YannikSc** [comment]
Can't you just revert from a fork? How long ago was the breach?

**EDIT**

With that I don't want to say, that you should fix it, its more like a thought, if someone wants to revive the project, and I'm rather asks for an estimation if this could be plausible way to get some sort of fix together.

Would be hard though if it was a couple months or even years ago.

**Comment by comfmai** [comment]
So I did a little digging:
This is the original binary version, for some reason it was compressed with UPX, which should have raised my suspicion (but I can see the previous versions are also UPXed, @lawl is this on purpose?) 

I put it through VirusTotal, it came out clean
https://www.virustotal.com/gui/file/726a3dd0b72d2de56b55d2920fd1ea64c5017983eb308f3e3bfba2dbe867ea94/detection

I also the unpacked version  through VirusTotal, it came out clean as well
https://www.virustotal.com/gui/file/23d531efde629161f64294a6f13a7c9ad7a6df06435d2fcc9f6604848959a2ac/detection

But, when firing WireShark I can see this:
TLS connection being made to `82.118.227.155` , which seems to be already reported: https://www.abuseipdb.com/check/82.118.227.155

This looks like a command & control server...

**Comment by lawl** [comment]
Yes upx was on purpose, but ypu shouldnt trust me when i say that because i already said my sytem(s?) were compromised.

**Comment by lawl** [comment]
82.118.227.155 have you checked this isnt the update server? Does noisetorch.epicgamer.org resolve there?

**Comment by Xunjin** [comment]
> Yes upx was on purpose, but ypu shouldnt trust me when i say that because i already said my sytem(s?) were compromised.

I see well I've looked for this IP and I found not a single reference using wireshark as @comfmai found out. However, I'm not an expert on this.

@lawl I know you are tired and feeling really depressed about it, however if there is a way for us to help you, maybe conduct some kind of audit. This project is way too good to die :cry: 

Edit: tho I'm using the arch repo https://aur.archlinux.org/packages/noisetorch, and it's indeed old.

**Comment by lawl** [comment]
Ok ive pointed the readme to this thread. If we can get enough people to check the code, maybe we can work from there.

**Comment by papaya2k** [comment]
Name:	noisetorch.epicgamer.org
Address: 82.118.227.155
Name:	noisetorch.epicgamer.org
Address: 2a01:8740:1:fe3f:dc78:593f:d16c:1

**Reply by Widward1** [reply]
That is the hostname of my install, I have already uninstalled it, but is there anything else that I should do?

**Comment by reynoldsme** [comment]
So, noisetorch.epicgamer.org does resolve to 82.118.227.155 but it also resolves to the ipv6 address 2a01:8740:1:fe3f:dc78:593f:d16c:1 so @Xunjin  you wouldn't necessarily see any references in wireshark to the 82.* address if testing from an ipv6 enabled network and machine.

**Comment by papaya2k** [comment]
> So, noisetorch.epicgamer.org does resolve to 82.118.227.155 but it also resolves to the ipv6 address 2a01:8740:1:fe3f:dc78:593f:d16c:1 so @Xunjin you wouldn't necessarily see any references in wireshark to the 82.* address if testing from an ipv6 enabled network and machine.

Don't know where others are from, but a lot of American ISPs including my own don't provide IPv6 access to the Internet.

**Comment by comfmai** [comment]
> Yes upx was on purpose, but ypu shouldnt trust me when i say that because i already said my sytem(s?) were compromised

I'm sorry my bad @lawl I should have double checked, I can see it here:
https://github.com/lawl/NoiseTorch/blob/2663bcc8d6aea1e15d0e52e47ce55f60283ef53b/Makefile#L22

> 82.118.227.155 have you checked this isnt the update server? Does noisetorch.epicgamer.org resolve there?

My double apology @lawl, Yes you are correct `noisetorch.epicgamer.org` is resolved to this address. I was confused because when I checked from a different machine I got a different IP, so I just assumed it was not the same one. Bottom line, this seems like a totally valid connection. @Xunjin this was the confusing part for me, if you try to resolve you would probably get a different IP (I'm assuming CDN in the middle)

I run WireShark and waited for several minutes, I could not detect any suspicious communication. So I think this is a good sign 🤞 

@lawl if there is anything I can help with, please feel free to suggest. NoiseTorch is awesome, you have earned my support and appreciation regardless of this unpleasant security breach. 

EDIT:
You can see here `noisetorch.epicgamer.org` is resolved to a different IP, anyhow I think this was false alarm on my end. apologies everyone 😓 
https://www.virustotal.com/gui/file/23d531efde629161f64294a6f13a7c9ad7a6df06435d2fcc9f6604848959a2ac/behavior

**Comment by Xunjin** [comment]
Thank you all for the info and feedbacks. Anyone with experience in audition? Would love tips on how to do it properly. :sweat_smile:

**Comment by reynoldsme** [comment]
> But, when firing WireShark I can see this: TLS connection being made to `82.118.227.155` , which seems to be already reported: https://www.abuseipdb.com/check/82.118.227.155
> 
> This looks like a command & control server...

I have no idea how reliable this AbuseIPDB is, but if accurate, it does seem to suggest that the host running noisetorch.epicgamer.org may be compromised (this assumes that 82.118.227.155 maps directly to the host and no NAT/proxy is involved). The requests in the reports look like a wordpress spam bot. That seems odd to be coming from a CDN address.

I see Caddy is serving requests which suggests this is a cloud vm instead of something like a shared web hosting platform?

**Comment by lawl** [comment]
Its a tiny vm with natted ipv4.

**Comment by Xunjin** [comment]
@lawl you pointed out in another issue these two commits https://github.com/lawl/NoiseTorch/commit/8c34658b64f1efeab501bef57d2bfa9579fe34e2 and https://github.com/lawl/NoiseTorch/commit/38787e4195f2a34d7ec4421caf17cb99bc31fa2b might be hiding something, I will skirmish through them. Any more commits that you believe might be a problem?

**Comment by 10maurycy10** [comment]
Sorry if this is hijacking this thread, but what was the exact scope of the compromize?

**Comment by lawl** [comment]
@Xunjin litterally all of them. These are just where i'd hide as an attacker, probably. As always absence of evidence != evidence of absence.

**Comment by Dadadah** [comment]
Rad, I'm glad this took off. Before I start taking a look at this, I wanna ask why the vendor folder is checked in to git. Usually with go mod you put the vendor folder in the .gitignore.

I'll do my part and try to comb through diffs and the code in general. While I'm not a security engineer, I do work in Go professionally.

**Comment by Dadadah** [comment]
My guess is that the biggest possible vector for attack is your update server. Perhaps we can disable that feature until we get everything looked over.

**Comment by Technetium1** [comment]
@10maurycy10 assume the worst until proven otherwise.

**Comment by Dadadah** [comment]
So just catching back up to this, I did a pretty in depth review of most of the stuff in this repo. Here's what I covered and as far as I can tell is working as expected:

Noisetorch requesting root for modifying capabilities
Noisetorch updating from a remote server (ASSUMING REMOTE SERVER IS TRUSTED)
Noisetorch changing the RLimit of PulseAudio
Noisetorch module loading and unloading
Noisetorch CLI
The last year of commit history for this repository

Some things I haven't checked:

Anything UI related
The PulseAudio library in use, which can be found at [lawl/pulseaudio](https://github.com/lawl/pulseaudio)
Untar code
Any C code

I am confident that when it comes to root access used by Noisetorch, the source is clean. However, do keep in mind that I am not a security engineer.

**Reply by ZyanKLee** [reply]
@Dadadah did you look into the `scripts/` folder as well?

**Reply by Dadadah** [reply]
Not when I did this. I can do so today along with the rest of the GO code from this repository.

**Comment by Dadadah** [comment]
While not a fix, #257 should make an audit much easier.

**Comment by MagicRB** [comment]
@lawl I've just learned of this from a nixpkgs PR and I'm so sorry. Security keys are one of the most private things a developer has. I'm not a golang developer so idk if I reasonably help. But good luck to everyone, especially you. 

PS. I know this is hard on you, but would you be able to talk about precisely what happened and how you lost your keys? Your experience could help others to not loose their keys. If it's too soon, it's fine, take your time.

**Comment by Xunjin** [comment]
@lawl I believe @Dadadah has a good point about the remote server, might help in reducing the surface of attack (however I'm not a security engineer, so take my point with a pinch of salt)

**Comment by MagicRB** [comment]
Ill just list what i have and know, so that if i can help, you people can tell me. I know Rust, Haskell, Nix, system admin stuff and I have 3 free VPSs rn in Oracle Cloud. IDK if any of that will be of use.

**Comment by principis** [comment]
I don't quite understand the need to check the entire source. Wouldn't it be good enough to check the latest commits? The key must have been compromised for a long time to become a major problem. If so, I suspect someone would have noticed it earlier?

**Comment by lawl** [comment]
yeah you can remove the remote server that's being pinged by people's builds by... shipping a new binary. otherwise, yeah the whole update code has to be ripped out at some point, but that's easy and not a problem.

**Comment by lawl** [comment]
> Wouldn't it be good enough to check the latest commits?

i'm now assuming it must likely have occured sometime in the second half 2021.

**Comment by lawl** [comment]
since someone asked, my assumption right now is that it was most likely just someone looking for bitcoin wallets. still not sure how they got in exactly, and now sure if we'll ever know, but as i already said, probably nothing related to noisetorch. and no bitcoin either.

still better safe than sorry.

**Comment by fuomag9** [comment]
82.118.227.155 is either a c&c server (lots of old and vulnerable services) because it has a lot of open ports with mainly ssh servers (plus some other stuff like NTP and two web servers) or a compromised host

Intestingly enough, `wq.steliosm.net` is a domain that resolved to that ip on the 15th of may.

`noisetorch.epicgamer.org
` has ssl certs [dating 2020] (https://search.censys.io/certificates/6358eed412e3ea1d13f0e3be08d943b2c2d56cae10ed3b88fe71c771a458528a), is this intended or you have been compromised from before? @lawl

**Comment by lawl** [comment]
@fuomag9 as i wrote somewhere else, this is an IPv4 NAT server with like 20 ports per VM, this is entirely expected.

**Comment by ZyanKLee** [comment]
Not that this is any proof for security, but I just let snyk scan my local clone of NoiseTorch and it came back with 0 issues on the dependencies.
Snyk does generate a pretty overview of the dependencies, though. Perhaps someone with a developers background can check the dependency list for sanity: https://paste.ubuntu.com/p/n69FMfmpGK/

As a next step I ran `gosec`, which did indeed find some issues - but on a first glance by me (a layman) none of those seem to be malicious in nature. I provide this as a paste for others to check it out, too. https://paste.ubuntu.com/p/HhxGmSSvvC/

**Comment by principis** [comment]
I went over a bunch of commits and looked at the code. I don't think it's tampered with. I will keep my builds for Fedora up, and keep using this awesome piece of software.

@lawl I'm sorry to hear that this happened to you, and I hope you'll recover soon!

**Comment by ghost** [comment]
Hello !

I went over the whole of the c/ directory and checked for any things that may be malicious and did NOT find anything to be suspicious. I spent about 2 hours looking over it so i'm fairly confident it is all okay. I am by no means a security researcher but program in C professionally and in my spare time.

@lawl I am really sorry to hear the news about all of this happening, just remember that we all have your back and owe you the world for the amazing piece of software you have written, we thank you greatly and hope you feel better soon.

**Comment by fuomag9** [comment]
> Not that this is any proof for security, but I just let snyk scan my local clone of NoiseTorch and it came back with 0 issues on the dependencies. Snyk does generate a pretty overview of the dependencies, though. Perhaps someone with a developers background can check the dependency list for sanity: https://paste.ubuntu.com/p/n69FMfmpGK/
> 
> As a next step I ran `gosec`, which did indeed find some issues - but on a first glance by me (a layman) none of those seem to be malicious in nature. I provide this as a paste for others to check it out, too. https://paste.ubuntu.com/p/HhxGmSSvvC/

I had not much time to check those, but that is indeed dangerous code from a security standpoint

An example, launching a process with os.environ passed is not a great idea per se:

```
syscall.Exec(self, []string{""}, os.Environ())
```

Nonetheless, those need to be put in context of when they were added and why so we can exclude them

**Comment by AXDOOMER** [comment]
@lawl Can you be more specific about how you found out you were compromised? How did it happen? Did you see any signs of an attacker? Do you have log files? Do you still have the latest binary from your server? (the one that may have been compromised)

I have a few old releases of noise torch. I can diff the binaries to find out if any malicious code was added. With the right tools, it's easy. 

Unless you're more specific about how it happened and how you found out, we can't know where to focus our efforts. Maybe this repository wasn't even compromised and only the update server got malicious payloads uploaded. I know this is depressing, but you must be transparent so we can make sure no further damage is caused. You'll feel relieved once we'll be able to pinpoint exactly a piece of code that's malicious, a version of the binary that's compromised or even find out that the project was not affected at all. 

----------------

What I believe can be done for now, assuming it wasn't only the update server that got compromised (anybody can participate): 

1- For every fork of the project, check if the commit history of this repo matches that of the fork. Also check if the content of each commit matches. 

2- For every dependency in the vendor folder, check if the code matches with the same version of the dependency in its original repo. The git repo of every dependency found in vendor/ can be found in go.mod. 

Document the versions, forks, dependencies that you've tested and your findings, if any.

**Comment by lawl** [comment]
Found a very suspicious process in htop. Paniced. Later straced it and it was looking for wallet.dat. The OS itself was fairly fresh (q3 ish?).

Sorry i dont think i have any logs or anything that isnt deleted. As you may see from my history, i paniced fairly hard.

**Comment by ntimo** [comment]
Did your saw the comment by THOR on the virus total community tab? https://www.virustotal.com/gui/file/726a3dd0b72d2de56b55d2920fd1ea64c5017983eb308f3e3bfba2dbe867ea94/community

**Comment by AXDOOMER** [comment]
> Did your saw the comment by THOR on the virus total community tab? https://www.virustotal.com/gui/file/726a3dd0b72d2de56b55d2920fd1ea64c5017983eb308f3e3bfba2dbe867ea94/community

Too generic of a YARA rule. It will basically detect anything packed with UPX. 
https://github.com/Neo23x0/signature-base/blob/master/yara/gen_elf_file_anomalies.yar

**Comment by ntimo** [comment]
I also scanned a older version of noisetorch from my previus ubuntu machine using virustotal and there mcaffee detects some generic Artemis!Trojan not sure how relevant this is.
https://www.virustotal.com/gui/file/a4ac5e60d0f3ac1a673fdcd997dbb22f6e2160d8f140afd2d435978a8fb1680a/detection
https://service.mcafee.com/?locale=en-US&articleId=TS100414&page=shell&shell=article-view

**Comment by AXDOOMER** [comment]
> I also scanned a older version of noisetorch from my previus ubuntu machine using virustotal and there mcaffee detects some generic Artemis!Trojan not sure how relevant this is. [link1](https://www.virustotal.com/gui/file/a4ac5e60d0f3ac1a673fdcd997dbb22f6e2160d8f140afd2d435978a8fb1680a/detection) [link2](https://service.mcafee.com/?locale=en-US&articleId=TS100414&page=shell&shell=article-view)

Not a big deal. This file was submitted to VT on 2021-10-01 for the first time and it was only detected by one AV, which means it's probably a false-positive.

**Comment by ZyanKLee** [comment]
@lawl perhaps pin this topic, so it can be easily found.

**Comment by protectroot-com** [comment]
Please post a link in your readme to https://www.buymeacoffee.com/ so we can help make some lemon-aid out of these lemons :)

**Comment by protectroot-com** [comment]
https://www.sandflysecurity.com/

**Comment by ntimo** [comment]
@protectroot-com did you used https://www.sandflysecurity.com/ to check your devices? Did it found something?

**Comment by protectroot-com** [comment]
I was only sharing a good resource.  I follow the owner of sandfly security on twitter.  He is the best security expert on Linux IMHO.  https://twitter.com/CraigHRowland

**Comment by TheDukeofErl** [comment]
> 2- For every dependency in the vendor folder, check if the code matches with the same version of the dependency in its original repo. The git repo of every dependency found in vendor/ can be found in go.mod.

Went and did a somewhat lazy way of doing this just by deleting the vendors folder and doing a `go mod vendor`, git diff returned that there were no differences by doing this so it's likely the vendor folder is fine. I didn't check the c/ folder though. It looks like that's more or less just vendor code (rnnoise and ladspa) so the same thing should be doable there.

One thing that's worth noting here is that the https://github.com/lawl/pulseaudio dependency may need some additional inspection, as that repo is also owned by lawl. It doesn't look like there's too much code there though, thankfully.

**Reply by ZyanKLee** [reply]
I verified the vendor folder with the same approach. :heavy_check_mark:

**Comment by Namenot** [comment]
Regarding the https://github.com/lawl/pulseaudio dependency, I doubt that it contains anything malicious (at least not from the current compromise) as there are only 2 commits in it from around q3, one was commited in June and the other in September. Both hardly change anything (less than a line of code each). Neither looking sus to me tbh. But I'm not a go dev so maybe I'm mistaken.

**Comment by AXDOOMER** [comment]
I went through https://github.com/lawl/pulseaudio twice and didn't see anything suspicious. I haven't yet finished going through NoiseTorch's source code. 

@lawl Are there reasons to think your Github was compromised? (password reuse between the server and Github, RSA private keys or SSH private keys on the server?) Did you use password authentification on the server? 

IMO NoiseTorch's update process should be changed so updates are pulled directly from Github releases rather than a server on the Internet.

**Comment by ZyanKLee** [comment]
Thanks @AXDOOMER for taking the time of looking through the code of the pulseaudio library.

I agree that the releases should be discovered and pulled from github releases. Let's put it on the list of things to be done in the next days/weeks.

**Comment by TheDukeofErl** [comment]
I've checked c/rnnoise against https://github.com/xiph/rnnoise, this dependency looks fine as well.

I know https://github.com/noisetorch/NoiseTorch/pull/257 was closed but I think it's worth revisiting. It would probably be good to investigate moving other dependencies, like rnnoise, out of the repo in a similar manner.

**Reply by TheDukeofErl** [reply]
Verified c/ringbuf.* against https://github.com/dhess/c-ringbuf.

**Reply by TheDukeofErl** [reply]
I've made a PR that removes both rnnoise and c-ringbuf as code and instead uses submodules: https://github.com/noisetorch/NoiseTorch/pull/268. Once the audit is done it can be tested then merged.

**Reply by OldiLo** [reply]
I made pr #269 for small fix and update dependencies

**Reply by TheDukeofErl** [reply]
I've finished verifying c/ladspa, it looks good to me.
```
.gitignore:	fine, verified manually
export.txt:	fine, verified manually
ladspa.h:	fine, verified against upstream (https://www.ladspa.org/ladspa_sdk/ladspa.h.txt)
Makefile:	fine, verified manually
module.c:   	fine, verified manually, based on ladspa_sdk/src/plugins/filter.c
utils.h:	fine, verified against upstream (https://www.ladspa.org/ladspa_sdk/overview.html)
```
This concludes my review of the c/ folder.

**Comment by ZyanKLee** [comment]
I will move this ticket to discussions for now. We are actually touching too many topics here for it to be an actual issue ticket.

**Comment by MoJo1760** [comment]
@lawl I'm not the best at code review, but I do things on the offensive side of security including chasing bad actors - can I help in any way?

**Comment by viniciusnevescosta** [comment]
Hello everyone, well, I'm a little lost about what happened, but my microphone started getting buggy (understand buggy as, popped and inaudible).

I was using noise torch normally but I decided to enter its repository, which contained some instructions to uninstall it, so I did. 

I don't know if these 2 facts correlate, but I would like to know why uninstall this program that until recently was normal. If you can explain to me what really happened, thank you!

**Reply by ZyanKLee** [reply]
I'm almost certain that this was a coincidence. Nevertheless: welcome!

**Reply by viniciusnevescosta** [reply]
Okay, but what really happened?

**Reply by viniciusnevescosta** [reply]
My mic is still buggy even though I uninstalled the program

**Reply by Dany-Wilde** [reply]
debian sid unstable . for me, everything works fine . 
where should I investigate ? what should I look at ? what tests should I do ? in what way ? with what command(s) ? if I can help ...
I'm a beginner in development. I don't know anything about it and, at my age, I don't understand a lot of things.

**Reply by viniciusnevescosta** [reply]
> debian sid unstable . for me, everything works fine . where should I investigate ? what should I look at ? what tests should I do ? in what way ? with what command(s) ? if I can help ... I'm a beginner in development. I don't know anything about it and, at my age, I don't understand a lot of things.

Hi, did you have any problems with respect to microphone or even privacy/security?

**Reply by ZyanKLee** [reply]
@Jolonte your microphone issues are not the result of the original author's PC having been hacked. If something broke and is still broken after you uninstalled NoiseTorch, you should search the issue elsewhere. For example with the drivers, pulseaudio/pipewire or your kernel.

**Reply by viniciusnevescosta** [reply]
> 

Yes, actually it has already been corrected. Via this issue: #273

**Comment by ghost** [comment]
Do we have infos about the update server? Was it compromised too? We might have to check it too (but no idea how) to be sure it didn't trigger fake updates.

**Reply by ZyanKLee** [reply]
Probably not, but we can't rule it out at this point. We are thinking about replacing it with GitHub releases. See #272 and #271

**Reply by ghost** [reply]
Yes, this is a good idea for the future. My question was more directed towards whether the server has been an attack vector, but as you say, we can't infallibly know for the moment.

**Comment by ZyanKLee** [comment]
I checked the Makefile and am pretty sure that in it itself there is no malicious code. Though I got two work packages out of the review: rework the update process with GH releases in #272 and #271

**Comment by Namenot** [comment]
I'm fairly inexperienced when it comes to github and code review, so excuse my naivety, but wouldn't it suffice to 'just' got through all commits done since the compromise instead of going through the whole code? Anyhow I still wanna help so if someone can point me in the right directions that would be nice as i have never done an audit myself and don't really know how to approach one.

**Reply by ZyanKLee** [reply]
given the right permissions, one could easily rewrite the whole git-history. Therefore I would deem it easier to review the current state and when that is in an apparent clean state, we can continue from there.
I'm not a security engineer myself, though, and one with that background could have another opinion on that matter.

As for how to help:
We got an overview of elements we need to check and read through at #275. If you feel you are somewhat experienced in any of these technologies, please feel free to do what you think is needed to verify its integrity.
After that leave a comment in #275, explaining what you did (so others can verify) and if you found anything suspicious.

**Comment by maggick** [comment]
I check the short Makefile. Nothing malicious there.
It just call `go run scripts/signer.go`.

Nonetheless it calls system binary (as `go`, `tar` or `upx`) if @lawl computer was fully compromised this system commands should not be trusted and so the binary should not be trusted. A new build should be fine once we reviewed the whole code.

**Reply by ZyanKLee** [reply]
Thank you for giving it a quick glance. I came to the same conclusion: it's too short and tidy to really hide something in there.

**Comment by ZyanKLee** [comment]
**Folks: YOU ARE AWESOME!
Thanks a lot for all the help with this code review. I'm glad you are here and helped out; I could not have done that without you (quite literally).**

Next goal is to create a new stable release with a working update mechanism #271 - perhaps someone wants to pick up that ticket and train those coding muscles a bit?

**Reply by ZyanKLee** [reply]
PS: I'll lock this topic soon, as we are done with the audit. If you wanted to add something to it afterwards, just open another discussion.

**Comment by ZyanKLee** [comment]
Due to a suspected security breach of the update server and code repository, there's
 been a concerted effort by the NoiseTorch community to ensure the source code and
 binaries are free from malicious code.
 
> No malicious code has been found.
 
 You can read more about the audit that was done [here](https://github.com/noisetorch/NoiseTorch/discussions/275)
 and [here](https://github.com/noisetorch/NoiseTorch/discussions/264).
 Updates will now be retrieved from the project's releases page to avoid any risk
 of this reoccurring. We thank everyone for their trust and the love that they've
 shown towards the project in this unpleasant time.
